File: android/src/main/java/com/strata/storage/SQLiteStorage.java
package com.strata.storage;

import android.content.Context;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;
import com.getcapacitor.JSObject;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * SQLite-backed storage. One {@link SQLiteStorage} instance maps to a single
 * database file (one {@link SQLiteOpenHelper}). A single instance can serve
 * multiple logical tables: every public method takes a sanitized {@code table}
 * name and the table is created on first use via {@code CREATE TABLE IF NOT
 * EXISTS}.
 *
 * <p>Value-shape contract (matches the TS {@code SqliteAdapter}): {@code set}
 * receives the FULL wrapper object ({@code value, created, updated, expires?,
 * tags?, metadata?}). The entire wrapper is JSON-serialized into the
 * {@code value} column for round-trip fidelity, while {@code created},
 * {@code updated}, {@code expires}, {@code tags} and {@code metadata} are also
 * extracted into their own columns for TTL + query support. {@code get} parses
 * the {@code value} column back into a {@link JSObject} wrapper.
 *
 * <p>All database access is synchronized on the instance so overlapping bridge
 * calls cannot corrupt the connection.
 */
public class SQLiteStorage extends SQLiteOpenHelper {
    private static final int DATABASE_VERSION = 1;
    private static final String DEFAULT_TABLE = "storage";

    private static final String KEY_ID = "key";
    private static final String KEY_VALUE = "value";
    private static final String KEY_CREATED = "created";
    private static final String KEY_UPDATED = "updated";
    private static final String KEY_EXPIRES = "expires";
    private static final String KEY_TAGS = "tags";
    private static final String KEY_METADATA = "metadata";

    /** Strict SQL-identifier allow-list for table names (cannot be bound). */
    private static final Pattern IDENTIFIER = Pattern.compile("^[A-Za-z_][A-Za-z0-9_]*$");

    /** Tables already created in this DB during the process lifetime. */
    private final Set<String> ensuredTables = new HashSet<>();

    /**
     * @param context  Android context
     * @param dbName    fully-resolved database file name (already sanitized +
     *                  suffixed with ".db" by the plugin)
     */
    public SQLiteStorage(Context context, String dbName) {
        super(context, dbName, null, DATABASE_VERSION);
    }

    /**
     * Validate a table name as a strict SQL identifier
     * ({@code ^[A-Za-z_][A-Za-z0-9_]*$}). Table names cannot be passed as bound
     * parameters in SQLite, so they MUST be validated before being interpolated
     * into DDL/DML. A null/empty name resolves to the default table; any other
     * non-conforming name is REJECTED ({@link IllegalArgumentException}) rather
     * than silently stripped — the TS adapter applies the same allow-list and
     * rejects before the call ever reaches here, so this is defense-in-depth
     * against a direct-plugin caller.
     */
    public static String sanitizeTable(String table) {
        if (table == null || table.isEmpty()) {
            return DEFAULT_TABLE;
        }
        if (!IDENTIFIER.matcher(table).matches()) {
            throw new IllegalArgumentException(
                    "Invalid SQLite table name: " + table
                    + " (must match ^[A-Za-z_][A-Za-z0-9_]*$)");
        }
        return table;
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        // Tables are created lazily per (database, table) via ensureTable().
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        // No destructive migrations: tables are created on demand and persist.
    }

    /**
     * Create the table for {@code safeTable} if it does not yet exist. Caller
     * MUST pass an already-sanitized identifier.
     */
    private void ensureTable(SQLiteDatabase db, String safeTable) {
        if (ensuredTables.contains(safeTable)) {
            return;
        }
        String createTable = "CREATE TABLE IF NOT EXISTS " + safeTable + "("
                + KEY_ID + " TEXT PRIMARY KEY,"
                + KEY_VALUE + " TEXT NOT NULL,"
                + KEY_CREATED + " INTEGER NOT NULL,"
                + KEY_UPDATED + " INTEGER NOT NULL,"
                + KEY_EXPIRES + " INTEGER,"
                + KEY_TAGS + " TEXT,"
                + KEY_METADATA + " TEXT" + ")";
        db.execSQL(createTable);
        ensuredTables.add(safeTable);
    }

    /**
     * Persist the full wrapper for {@code key}. The wrapper is stored verbatim
     * as JSON in the {@code value} column; {@code created}/{@code updated}/
     * {@code expires}/{@code tags}/{@code metadata} are mirrored into columns
     * for TTL + query.
     *
     * @param wrapper the full {@link StorageValue} wrapper as a {@link JSObject}
     */
    public synchronized boolean set(String table, String key, JSObject wrapper) {
        if (wrapper == null) {
            return false;
        }
        String safeTable = sanitizeTable(table);
        long now = System.currentTimeMillis();

        // Column extraction for TTL + query. Defaults keep legacy fidelity.
        long created = wrapper.optLong(KEY_CREATED, now);
        long updated = wrapper.optLong(KEY_UPDATED, now);

        ContentValues values = new ContentValues();
        values.put(KEY_ID, key);
        values.put(KEY_VALUE, wrapper.toString());
        values.put(KEY_CREATED, created);
        values.put(KEY_UPDATED, updated);

        if (wrapper.has(KEY_EXPIRES) && !wrapper.isNull(KEY_EXPIRES)) {
            values.put(KEY_EXPIRES, wrapper.optLong(KEY_EXPIRES));
        }
        // tags / metadata are stored as their JSON text for query use.
        Object tags = wrapper.opt(KEY_TAGS);
        if (tags != null && tags != JSONObject.NULL) {
            values.put(KEY_TAGS, tags.toString());
        }
        Object metadata = wrapper.opt(KEY_METADATA);
        if (metadata != null && metadata != JSONObject.NULL) {
            values.put(KEY_METADATA, metadata.toString());
        }

        try {
            SQLiteDatabase db = getWritableDatabase();
            ensureTable(db, safeTable);
            long result = db.insertWithOnConflict(
                    safeTable, null, values, SQLiteDatabase.CONFLICT_REPLACE);
            return result != -1;
        } catch (Exception e) {
            Log.e("StrataStorage", "Failed to set value in SQLite", e);
            return false;
        }
    }

    /**
     * Read the full wrapper for {@code key}, parsed back from the JSON stored in
     * the {@code value} column. Missing row → {@code null}. Unparseable legacy
     * bytes → {@code null} (treated as a miss; never throws).
     */
    public synchronized JSObject get(String table, String key) {
        String safeTable = sanitizeTable(table);
        SQLiteDatabase db = getReadableDatabase();
        ensureTable(db, safeTable);
        Cursor cursor = null;
        try {
            cursor = db.query(safeTable, new String[]{KEY_VALUE}, KEY_ID + "=?",
                    new String[]{key}, null, null, null, null);
            if (cursor != null && cursor.moveToFirst()) {
                String stored = cursor.getString(0);
                if (stored == null) {
                    return null;
                }
                try {
                    return new JSObject(stored);
                } catch (JSONException parseError) {
                    // Legacy / corrupted payload: treat as a miss rather than throw.
                    Log.w("StrataStorage", "Unparseable SQLite value for key " + key);
                    return null;
                }
            }
            return null;
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
    }

    public synchronized boolean remove(String table, String key) {
        String safeTable = sanitizeTable(table);
        SQLiteDatabase db = getWritableDatabase();
        ensureTable(db, safeTable);
        int result = db.delete(safeTable, KEY_ID + " = ?", new String[]{key});
        return result > 0;
    }

    public synchronized boolean clear(String table, String prefix) {
        String safeTable = sanitizeTable(table);
        SQLiteDatabase db = getWritableDatabase();
        ensureTable(db, safeTable);
        int result;
        if (prefix != null) {
            result = db.delete(safeTable, KEY_ID + " LIKE ?", new String[]{prefix + "%"});
        } else {
            result = db.delete(safeTable, null, null);
        }
        return result >= 0;
    }

    public synchronized List<String> keys(String table, String pattern) {
        String safeTable = sanitizeTable(table);
        long now = System.currentTimeMillis();
        List<String> keys = new ArrayList<>();
        SQLiteDatabase db = getReadableDatabase();
        ensureTable(db, safeTable);
        Cursor cursor = null;
        // Exclude expired rows in SQL so the TS adapter does not need a per-key
        // get() to drop them (eliminates the keys() N+1).
        String notExpired = "(" + KEY_EXPIRES + " IS NULL OR " + KEY_EXPIRES + " > ?)";
        String nowArg = String.valueOf(now);
        try {
            if (pattern != null) {
                cursor = db.query(safeTable, new String[]{KEY_ID},
                        KEY_ID + " LIKE ? AND " + notExpired,
                        new String[]{"%" + pattern + "%", nowArg}, null, null, null, null);
            } else {
                cursor = db.query(safeTable, new String[]{KEY_ID},
                        notExpired, new String[]{nowArg}, null, null, null, null);
            }
            if (cursor != null && cursor.moveToFirst()) {
                do {
                    keys.add(cursor.getString(0));
                } while (cursor.moveToNext());
            }
            return keys;
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
    }

    public synchronized boolean has(String table, String key) {
        String safeTable = sanitizeTable(table);
        SQLiteDatabase db = getReadableDatabase();
        ensureTable(db, safeTable);
        Cursor cursor = null;
        try {
            cursor = db.query(safeTable, new String[]{KEY_ID}, KEY_ID + "=?",
                    new String[]{key}, null, null, null, null);
            return cursor != null && cursor.getCount() > 0;
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
    }

    /**
     * Enumerate non-expired rows as {@code { key, value }} where {@code value}
     * is the full {@link StorageValue} wrapper (parsed from the stored JSON).
     * Returning the wrapper here lets the TS {@code SqliteAdapter} run the real
     * query condition in JS from a SINGLE round-trip instead of re-fetching
     * each key (eliminates the query() N+1). Expired rows are filtered in SQL;
     * an unparseable legacy payload surfaces as key-only (the TS side then
     * falls back to {@code get()}, which treats it as a miss).
     */
    public synchronized List<Map<String, Object>> query(String table) {
        String safeTable = sanitizeTable(table);
        long now = System.currentTimeMillis();
        List<Map<String, Object>> rows = new ArrayList<>();
        SQLiteDatabase db = getReadableDatabase();
        ensureTable(db, safeTable);
        Cursor cursor = null;
        try {
            cursor = db.query(safeTable, new String[]{KEY_ID, KEY_VALUE},
                    "(" + KEY_EXPIRES + " IS NULL OR " + KEY_EXPIRES + " > ?)",
                    new String[]{String.valueOf(now)}, null, null, null, null);
            if (cursor != null && cursor.moveToFirst()) {
                do {
                    Map<String, Object> row = new HashMap<>();
                    row.put("key", cursor.getString(0));
                    String stored = cursor.getString(1);
                    if (stored != null) {
                        try {
                            row.put("value", new JSObject(stored));
                        } catch (JSONException parseError) {
                            // Unparseable legacy payload: surface key only.
                            Log.w("StrataStorage", "Unparseable SQLite value in query for key "
                                    + cursor.getString(0));
                        }
                    }
                    rows.add(row);
                } while (cursor.moveToNext());
            }
            return rows;
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
    }

    /**
     * Delete every expired row ({@code expires <= now}) in one statement and
     * return the number removed. Because {@link #keys} and {@link #query} now
     * filter expired rows in SQL (no lazy per-key deletion on read), this is
     * what reclaims their physical storage; the TS adapter calls it on its TTL
     * cleanup tick.
     */
    public synchronized int cleanupExpired(String table) {
        String safeTable = sanitizeTable(table);
        long now = System.currentTimeMillis();
        SQLiteDatabase db = getWritableDatabase();
        ensureTable(db, safeTable);
        return db.delete(safeTable,
                KEY_EXPIRES + " IS NOT NULL AND " + KEY_EXPIRES + " <= ?",
                new String[]{String.valueOf(now)});
    }

    /**
     * Size information. {@code total} = Σ(length(key) + length(value)), matching
     * the web adapters' convention (key bytes are included). When {@code
     * detailed} is true the byte breakdown is also populated: {@code keys} = Σ
     * length(key), {@code values} = Σ length(value), {@code metadata} = Σ
     * length(metadata) (0 for null); {@code count} is the row count.
     */
    public synchronized SizeInfo size(String table, boolean detailed) {
        String safeTable = sanitizeTable(table);
        SQLiteDatabase db = getReadableDatabase();
        ensureTable(db, safeTable);
        Cursor cursor = null;
        try {
            if (!detailed) {
                cursor = db.rawQuery("SELECT COUNT(*), SUM(LENGTH(" + KEY_ID + ") + LENGTH("
                        + KEY_VALUE + ")) FROM " + safeTable, null);
                long totalSize = 0;
                int count = 0;
                if (cursor != null && cursor.moveToFirst()) {
                    count = cursor.getInt(0);
                    if (!cursor.isNull(1)) {
                        totalSize = cursor.getLong(1);
                    }
                }
                return new SizeInfo(totalSize, count);
            }

            cursor = db.query(safeTable,
                    new String[]{KEY_ID, KEY_VALUE, KEY_METADATA},
                    null, null, null, null, null);
            long keysBytes = 0;
            long valuesBytes = 0;
            long metadataBytes = 0;
            int count = 0;
            if (cursor != null && cursor.moveToFirst()) {
                do {
                    count++;
                    String key = cursor.getString(0);
                    String value = cursor.getString(1);
                    String metadata = cursor.getString(2);
                    if (key != null) {
                        keysBytes += key.getBytes(StandardCharsets.UTF_8).length;
                    }
                    if (value != null) {
                        valuesBytes += value.getBytes(StandardCharsets.UTF_8).length;
                    }
                    if (metadata != null) {
                        metadataBytes += metadata.getBytes(StandardCharsets.UTF_8).length;
                    }
                } while (cursor.moveToNext());
            }
            return new SizeInfo(keysBytes + valuesBytes, count, keysBytes, valuesBytes, metadataBytes);
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
    }

    /**
     * Size information. The detailed byte breakdown is only meaningful when
     * {@link #detailed} is true.
     */
    public static class SizeInfo {
        public final long total;
        public final int count;
        public final boolean detailed;
        public final long keysBytes;
        public final long valuesBytes;
        public final long metadataBytes;

        public SizeInfo(long total, int count) {
            this.total = total;
            this.count = count;
            this.detailed = false;
            this.keysBytes = 0;
            this.valuesBytes = 0;
            this.metadataBytes = 0;
        }

        public SizeInfo(long total, int count, long keysBytes, long valuesBytes, long metadataBytes) {
            this.total = total;
            this.count = count;
            this.detailed = true;
            this.keysBytes = keysBytes;
            this.valuesBytes = valuesBytes;
            this.metadataBytes = metadataBytes;
        }
    }
}

File: ios/Plugin/SQLiteStorage.swift
import Foundation
import SQLite3
import os.log

/// SQLITE_TRANSIENT tells SQLite to copy bound bytes immediately, which is
/// required when binding Swift `String`/`Data` whose buffers may be freed
/// before `sqlite3_step` runs. Passing `nil` (SQLITE_STATIC) here is a
/// use-after-free hazard for transient Swift buffers.
private let SQLITE_TRANSIENT = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

/**
 * SQLite-backed storage for one `(database, table)` pair.
 *
 * Multi-store design:
 * - One `SQLiteStorage` instance owns one open DB handle for one database file.
 * - A single instance can serve many tables in that file; tables are created
 *   on first use (see `ensureTable`).
 * - The plugin caches one instance per database filename (see
 *   `SQLiteStorageManager`) so the DB is opened once, not per call.
 *
 * Value-shape contract (matches src/plugin/definitions.ts):
 * - `set` receives the FULL `StorageValue` wrapper object
 *   `{ value, created, updated, expires?, tags?, metadata?, ... }`. The whole
 *   wrapper is JSON-serialized into the `value` column for a perfect
 *   round-trip. `created`/`updated`/`expires`/`tags`/`metadata` are ALSO
 *   mirrored into dedicated columns so native TTL cleanup + `query` can use
 *   them without re-parsing the blob.
 * - `get` parses the `value` column bytes back into the wrapper object and the
 *   plugin resolves `{ value: wrapper }`. Corrupt/legacy bytes that do not
 *   parse as a JSON object are treated as a miss (never a crash).
 *
 * SQL-safety: only the table identifier is interpolated into SQL, and it is
 * pre-sanitized to `^[A-Za-z0-9_]+$` by the plugin. All user values are bound
 * parameters.
 */
@objc public class SQLiteStorage: NSObject {
    private var db: OpaquePointer?
    private let dbName: String
    /// Tables already verified/created on this handle, to skip redundant DDL.
    private var ensuredTables: Set<String> = []
    private let logger = OSLog(subsystem: "com.strata.storage", category: "SQLiteStorage")

    /// Opens (or creates) the database file `<dbName>` under Documents.
    /// `dbName` MUST already be a sanitized filename (e.g. `storage.db`).
    @objc public init(dbName: String = "strata_storage.db") {
        self.dbName = dbName
        super.init()
        do {
            try openDatabase()
        } catch {
            os_log("Failed to initialize SQLite storage: %{public}@", log: logger, type: .error, error.localizedDescription)
        }
    }

    @objc public var isOpen: Bool {
        return db != nil
    }

    deinit {
        closeDatabase()
    }

    private func openDatabase() throws {
        guard let fileURL = try? FileManager.default
            .url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: false)
            .appendingPathComponent(dbName) else {
            throw NSError(
                domain: "StrataStorage.SQLiteStorage",
                code: 1001,
                userInfo: [NSLocalizedDescriptionKey: "Unable to access document directory"]
            )
        }

        guard sqlite3_open(fileURL.path, &db) == SQLITE_OK else {
            let errorMessage = String(cString: sqlite3_errmsg(db))
            sqlite3_close(db)
            db = nil
            throw NSError(
                domain: "StrataStorage.SQLiteStorage",
                code: 1002,
                userInfo: [NSLocalizedDescriptionKey: "Unable to open database: \(errorMessage)"]
            )
        }
    }

    private func closeDatabase() {
        if db != nil {
            sqlite3_close(db)
            db = nil
        }
    }

    /// Creates the table for `table` if it does not exist yet. Cached per
    /// instance so repeated calls are cheap. `table` MUST be pre-sanitized.
    private func ensureTable(_ table: String) throws {
        guard let db = db else {
            throw notInitializedError()
        }
        if ensuredTables.contains(table) { return }

        let createTableString = """
            CREATE TABLE IF NOT EXISTS \(table) (
                key TEXT PRIMARY KEY NOT NULL,
                value BLOB NOT NULL,
                created INTEGER NOT NULL,
                updated INTEGER NOT NULL,
                expires INTEGER,
                tags TEXT,
                metadata TEXT
            );
        """

        var createTableStatement: OpaquePointer?
        defer { sqlite3_finalize(createTableStatement) }

        guard sqlite3_prepare_v2(db, createTableString, -1, &createTableStatement, nil) == SQLITE_OK else {
            let errorMessage = String(cString: sqlite3_errmsg(db))
            throw NSError(
                domain: "StrataStorage.SQLiteStorage",
                code: 1003,
                userInfo: [NSLocalizedDescriptionKey: "Failed to prepare CREATE TABLE: \(errorMessage)"]
            )
        }

        guard sqlite3_step(createTableStatement) == SQLITE_DONE else {
            let errorMessage = String(cString: sqlite3_errmsg(db))
            throw NSError(
                domain: "StrataStorage.SQLiteStorage",
                code: 1004,
                userInfo: [NSLocalizedDescriptionKey: "Failed to create table: \(errorMessage)"]
            )
        }

        ensuredTables.insert(table)
    }

    private func notInitializedError() -> NSError {
        return NSError(
            domain: "StrataStorage.SQLiteStorage",
            code: 1000,
            userInfo: [NSLocalizedDescriptionKey: "Database not initialized"]
        )
    }

    /**
     * Persist a full `StorageValue` wrapper for `key` in `table`.
     *
     * The entire `wrapper` dictionary is JSON-encoded into the `value` column
     * (perfect round-trip). `created`/`updated`/`expires` and the JSON of
     * `tags`/`metadata` are extracted into their own columns for TTL + query.
     * Values are bound parameters; only `table` is interpolated (pre-sanitized).
     */
    @objc public func set(table: String, key: String, wrapper: [String: Any]) throws -> Bool {
        guard let db = db else { throw notInitializedError() }
        try ensureTable(table)

        // Serialize the whole wrapper for round-trip storage.
        let blob = try JSONSerialization.data(withJSONObject: wrapper, options: [])

        // Extract mirror columns from the wrapper. created/updated are required
        // by the wrapper contract; fall back to "now" defensively.
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let created = (wrapper["created"] as? NSNumber)?.int64Value ?? now
        let updated = (wrapper["updated"] as? NSNumber)?.int64Value ?? now
        let expires = (wrapper["expires"] as? NSNumber)?.int64Value

        let tagsData: Data?
        if let tags = wrapper["tags"] {
            tagsData = try? JSONSerialization.data(withJSONObject: tags, options: [])
        } else {
            tagsData = nil
        }

        let metadataData: Data?
        if let metadata = wrapper["metadata"] {
            metadataData = try? JSONSerialization.data(withJSONObject: metadata, options: [])
        } else {
            metadataData = nil
        }

        let insertSQL = """
            INSERT OR REPLACE INTO \(table)
            (key, value, created, updated, expires, tags, metadata)
            VALUES (?, ?, ?, ?, ?, ?, ?)
        """

        var statement: OpaquePointer?
        defer { sqlite3_finalize(statement) }

        guard sqlite3_prepare_v2(db, insertSQL, -1, &statement, nil) == SQLITE_OK else {
            return false
        }

        let ok =
            sqlite3_bind_text(statement, 1, key, -1, SQLITE_TRANSIENT) == SQLITE_OK &&
            blob.withUnsafeBytes { sqlite3_bind_blob(statement, 2, $0.baseAddress, Int32(blob.count), SQLITE_TRANSIENT) } == SQLITE_OK &&
            sqlite3_bind_int64(statement, 3, created) == SQLITE_OK &&
            sqlite3_bind_int64(statement, 4, updated) == SQLITE_OK &&
            (expires != nil ? sqlite3_bind_int64(statement, 5, expires!) : sqlite3_bind_null(statement, 5)) == SQLITE_OK &&
            bindOptionalBlob(statement, 6, tagsData) == SQLITE_OK &&
            bindOptionalBlob(statement, 7, metadataData) == SQLITE_OK

        guard ok else { return false }
        return sqlite3_step(statement) == SQLITE_DONE
    }

    /// Binds `data` as a blob to `index`, or NULL when `data` is nil.
    private func bindOptionalBlob(_ statement: OpaquePointer?, _ index: Int32, _ data: Data?) -> Int32 {
        guard let data = data else {
            return sqlite3_bind_null(statement, index)
        }
        return data.withUnsafeBytes {
            sqlite3_bind_blob(statement, index, $0.baseAddress, Int32(data.count), SQLITE_TRANSIENT)
        }
    }

    /**
     * Read the wrapper object for `key` from `table`.
     *
     * Returns the decoded `StorageValue` wrapper dictionary, or `nil` for a
     * missing row OR when stored bytes do not parse as a JSON object
     * (legacy/corrupt data is treated as a miss, never a crash).
     */
    @objc public func get(table: String, key: String) -> [String: Any]? {
        guard let db = db else {
            os_log("Database not initialized", log: logger, type: .error)
            return nil
        }
        // Reading a table that was never created is a miss, not an error.
        try? ensureTable(table)

        let querySQL = "SELECT value FROM \(table) WHERE key = ? LIMIT 1"
        var statement: OpaquePointer?
        defer { sqlite3_finalize(statement) }

        guard sqlite3_prepare_v2(db, querySQL, -1, &statement, nil) == SQLITE_OK,
              sqlite3_bind_text(statement, 1, key, -1, SQLITE_TRANSIENT) == SQLITE_OK else {
            return nil
        }

        guard sqlite3_step(statement) == SQLITE_ROW,
              let blob = sqlite3_column_blob(statement, 0) else {
            return nil
        }

        let valueData = Data(bytes: blob, count: Int(sqlite3_column_bytes(statement, 0)))
        // The wrapper is always a JSON object. Anything else = legacy/corrupt → miss.
        guard let wrapper = try? JSONSerialization.jsonObject(with: valueData, options: []) as? [String: Any] else {
            return nil
        }
        return wrapper
    }

    @objc public func remove(table: String, key: String) -> Bool {
        guard let db = db else {
            os_log("Database not initialized", log: logger, type: .error)
            return false
        }
        try? ensureTable(table)

        let deleteSQL = "DELETE FROM \(table) WHERE key = ?"
        var statement: OpaquePointer?
        defer { sqlite3_finalize(statement) }

        guard sqlite3_prepare_v2(db, deleteSQL, -1, &statement, nil) == SQLITE_OK,
              sqlite3_bind_text(statement, 1, key, -1, SQLITE_TRANSIENT) == SQLITE_OK else {
            return false
        }
        return sqlite3_step(statement) == SQLITE_DONE
    }

    @objc public func clear(table: String, prefix: String? = nil) -> Bool {
        guard let db = db else {
            os_log("Database not initialized", log: logger, type: .error)
            return false
        }
        try? ensureTable(table)

        let deleteSQL: String
        if prefix != nil {
            deleteSQL = "DELETE FROM \(table) WHERE key LIKE ?"
        } else {
            deleteSQL = "DELETE FROM \(table)"
        }

        var statement: OpaquePointer?
        defer { sqlite3_finalize(statement) }

        guard sqlite3_prepare_v2(db, deleteSQL, -1, &statement, nil) == SQLITE_OK else {
            return false
        }
        if let prefix = prefix {
            guard sqlite3_bind_text(statement, 1, "\(prefix)%", -1, SQLITE_TRANSIENT) == SQLITE_OK else {
                return false
            }
        }
        return sqlite3_step(statement) == SQLITE_DONE
    }

    @objc public func keys(table: String, pattern: String? = nil) -> [String] {
        guard let db = db else {
            os_log("Database not initialized", log: logger, type: .error)
            return []
        }
        try? ensureTable(table)

        // Exclude expired rows in SQL so the TS adapter does not need a per-key
        // get() to drop them (eliminates the keys() N+1).
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let notExpired = "(expires IS NULL OR expires > ?)"
        let querySQL: String
        if pattern != nil {
            querySQL = "SELECT key FROM \(table) WHERE key LIKE ? AND \(notExpired)"
        } else {
            querySQL = "SELECT key FROM \(table) WHERE \(notExpired)"
        }

        var statement: OpaquePointer?
        defer { sqlite3_finalize(statement) }
        var keys: [String] = []

        if sqlite3_prepare_v2(db, querySQL, -1, &statement, nil) == SQLITE_OK {
            var bindIndex: Int32 = 1
            if let pattern = pattern {
                // `%pattern%` mirrors the contains-matching used by other backends.
                sqlite3_bind_text(statement, bindIndex, "%\(pattern)%", -1, SQLITE_TRANSIENT)
                bindIndex += 1
            }
            sqlite3_bind_int64(statement, bindIndex, now)
            while sqlite3_step(statement) == SQLITE_ROW {
                if let key = sqlite3_column_text(statement, 0) {
                    keys.append(String(cString: key))
                }
            }
        }
        return keys
    }

    /**
     * Aggregate sizes for `table`.
     *
     * Returns `count`, `total` (= `keys` + `values`), and per-segment byte
     * sums: `keys` = SUM(LENGTH(key)), `values` = SUM(LENGTH(value)),
     * `metadata` = SUM(LENGTH(metadata)) treating NULL as 0. The plugin decides
     * whether to surface the `detailed` segment based on the `detailed` flag.
     */
    @objc public func size(table: String) throws -> [String: Int] {
        guard let db = db else { throw notInitializedError() }
        try ensureTable(table)

        let querySQL = """
            SELECT COUNT(*),
                   COALESCE(SUM(LENGTH(key)), 0),
                   COALESCE(SUM(LENGTH(value)), 0),
                   COALESCE(SUM(LENGTH(metadata)), 0)
            FROM \(table)
        """
        var statement: OpaquePointer?
        defer { sqlite3_finalize(statement) }

        var count = 0
        var keysSize = 0
        var valuesSize = 0
        var metadataSize = 0

        if sqlite3_prepare_v2(db, querySQL, -1, &statement, nil) == SQLITE_OK,
           sqlite3_step(statement) == SQLITE_ROW {
            count = Int(sqlite3_column_int(statement, 0))
            keysSize = Int(sqlite3_column_int64(statement, 1))
            valuesSize = Int(sqlite3_column_int64(statement, 2))
            metadataSize = Int(sqlite3_column_int64(statement, 3))
        }

        return [
            "count": count,
            "keys": keysSize,
            "values": valuesSize,
            "metadata": metadataSize,
            "total": keysSize + valuesSize
        ]
    }

    /**
     * Returns non-expired rows as `{ key, value }` where `value` is the full
     * `StorageValue` wrapper (parsed from the stored JSON blob). Returning the
     * wrapper here lets the JS SqliteAdapter run the real query `condition` in
     * JS from a SINGLE round-trip instead of re-fetching each key (eliminates
     * the query() N+1). Expired rows are filtered in SQL; an unparseable legacy
     * payload surfaces as key-only (the JS side falls back to `get()`, a miss).
     * The `condition` argument is accepted for forward-compatibility but is not
     * yet pushed down into SQL.
     */
    @objc public func query(table: String, condition: [String: Any]) -> [[String: Any]] {
        guard let db = db else {
            os_log("Database not initialized", log: logger, type: .error)
            return []
        }
        try? ensureTable(table)

        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let querySQL = "SELECT key, value FROM \(table) WHERE (expires IS NULL OR expires > ?)"
        var statement: OpaquePointer?
        defer { sqlite3_finalize(statement) }
        var rows: [[String: Any]] = []

        if sqlite3_prepare_v2(db, querySQL, -1, &statement, nil) == SQLITE_OK {
            sqlite3_bind_int64(statement, 1, now)
            while sqlite3_step(statement) == SQLITE_ROW {
                guard let keyCString = sqlite3_column_text(statement, 0) else { continue }
                var row: [String: Any] = ["key": String(cString: keyCString)]
                if let blob = sqlite3_column_blob(statement, 1) {
                    let valueData = Data(bytes: blob, count: Int(sqlite3_column_bytes(statement, 1)))
                    if let wrapper = try? JSONSerialization.jsonObject(with: valueData, options: []) as? [String: Any] {
                        row["value"] = wrapper
                    }
                }
                rows.append(row)
            }
        }
        return rows
    }

    /**
     * Delete every expired row (`expires <= now`) in one statement and return
     * the number removed. Because `keys` and `query` now filter expired rows in
     * SQL (no lazy per-key deletion on read), this is what reclaims their
     * physical storage; the JS adapter calls it on its TTL cleanup tick.
     */
    @objc public func cleanupExpired(table: String) -> Int {
        guard let db = db else {
            os_log("Database not initialized", log: logger, type: .error)
            return 0
        }
        try? ensureTable(table)

        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let deleteSQL = "DELETE FROM \(table) WHERE expires IS NOT NULL AND expires <= ?"
        var statement: OpaquePointer?
        defer { sqlite3_finalize(statement) }

        guard sqlite3_prepare_v2(db, deleteSQL, -1, &statement, nil) == SQLITE_OK,
              sqlite3_bind_int64(statement, 1, now) == SQLITE_OK,
              sqlite3_step(statement) == SQLITE_DONE else {
            return 0
        }
        return Int(sqlite3_changes(db))
    }
}

/**
 * Caches one open `SQLiteStorage` (and thus one DB handle) per database
 * filename, so the database is opened once per file instead of on every
 * bridge call. Access is serialized with a lock because Capacitor may dispatch
 * plugin calls on a background queue.
 */
final class SQLiteStorageManager {
    static let shared = SQLiteStorageManager()

    private var stores: [String: SQLiteStorage] = [:]
    private let lock = NSLock()

    private init() {}

    /// Returns the cached store for `fileName` (e.g. `storage.db`), creating it
    /// on first use. `fileName` MUST already be sanitized.
    func store(forFile fileName: String) -> SQLiteStorage {
        lock.lock()
        defer { lock.unlock() }
        if let existing = stores[fileName] {
            return existing
        }
        let created = SQLiteStorage(dbName: fileName)
        stores[fileName] = created
        return created
    }
}
